### DISCUSSION TITLE
Using matcher to find any token that is not an entity

### DISCUSSION THREAD

**Discussion by farrandi** [explosion/spaCy#8724] [Help: Coding & Implementations] [answered]
So I would like to create a matcher pattern that would be able to find this sentence structure (and similar structures):
![image](https://user-images.githubusercontent.com/54917098/125701916-a6e3c112-fa41-4dc9-b885-79bf97a352e6.png)

Currently, my pattern is:
```
pattern1 = [{"ENT_TYPE": "ANTMCL_STR", "OP": "+"}, {"ORTH": "is", "OP": "*"},
                   {"ENT_TYPE": "VALUE", "OP": "+"}, {"ORTH": "in", "OP": "*"},
                   {"ENT_TYPE": "OBS", "OP": "+"}]
matcher.add("pattern1", [pattern1], greedy = 'LONGEST')
```

I was wondering if I could find patterns without hardcoding the "is" and "in". Instead, I would like any characters even: "=", ":", "expected as", etc to be be included in between.

I tried using nothing and just {"OP": "*"} but it would just take a huge chunk of the text (entities included). 

How do I include any wildcard characters/words in between these entities that as long as it is not an entity, I would like the matcher to find it.

**Comment by polm** [comment, marked as answer]
To match a non-entity token you can use `{"ENT_TYPE":""}`. 

That said, in this case I think you might be better off changing your approach to one of these:

1. Use the dependency matcher instead.
2. Make several basic patterns (rather than one complex pattern) with the matcher like `ANTMCL is VALUE in OBS` and see how much coverage they get you
3. Split your input into sentences and put all the ents from each sentence in a dict to build output (oops, I guess that's what I suggested in https://github.com/explosion/spaCy/discussions/8635 too)

**Reply by farrandi** [reply]
I would like to try your "separate by sentences and use dict" suggestion however the sentencizer in spacy is not exactly separating the sentences well:
```
text = '''
Left Ventricle:   Normal
IVSd: *12 mm  (6-10) 
LVPWd: 9 mm  (6-10) 
LVIDd: 52 mm  (42-58) 
LVIDd index: 22 mm/m²  (22-30) 
LVIDs: 45 mm (25-40) 
LV EF (Visual): 57 % (52-72) 
LV Mass index: 102.5 g/m²  (49-115) 
LV RWT: 0.34 (<0.43)
LVOT diam: 23 mm  
SV LVOT: 61.7 ml  '''
text = re.sub(r'\s+',' ', text)
doc = nlp(text)
for sent in doc.sents:
    print(sent.text)
    print("---")
```
would give:
```
 Left Ventricle: Normal IVSd: *12 mm (6-10) LVPWd: 9 mm (6-10) LVIDd: 52 mm (42-58) LVIDd index: 22 mm/m² (22-30) LVIDs: 45 mm (25-40)
---
LV EF (Visual): 57 % (52-72) LV Mass index: 102.5 g/m² (49-115)
---
LV RWT: 0.34 (<0.43)
---
LVOT diam: 23 mm SV LVOT:
---
61.7 ml
---
```

I would like to separate the sentences so that it would separate each new line like that as a new sentence. Although in the text file sometimes there are sentences like: `Left Ventricle:  Normal size based on linear \nindex dimension.   Normal systolic function.` (there is a `\n` character before index dimension) This is why I subbed out all the hidden characters with a whitespace with regex before putting it into the doc object.

How would I be able to customize the tokenizer for this?
I read the documentation and I was interested in the options:
1. Use the sentencizer, with `punct_chars`. I tried adding `\n` as a punct_char and removed the regex sub but it did not work:
```
nlp = spacy.load("models/blank6/new-model")
nlp.remove_pipe("parser")
config = {"punct_chars": ['!','.','?','+',' \n']}
nlp.add_pipe("sentencizer", config=config)
#added same text here
doc = nlp(text)
for sent in doc.sents:
    print(sent.text)
    print("---")
```
it gave:
```
Left Ventricle:   Normal
IVSd: *12 mm  (6-10) 
LVPWd: 9 mm  (6-10) 
LVIDd: 52 mm  (42-58) 
LVIDd index: 22 mm/m²  (22-30) 
LVIDs: 45 mm (25-40) 
LV EF (Visual): 57 % (52-72) 
LV Mass index: 102.5 g/m²  (49-115) 
LV RWT: 0.34 (<0.43)
LVOT diam: 23 mm  

---
SV LVOT: 61.7 ml  
---
```

2. Using Custom rule-based strategy, which is using the language.component. Would this be better? I have not tried this.

**Reply by polm** [reply]
Ah, I didn't realize your text looked like that. Keep in mind that spaCy's training data is basically news articles, and so if your text isn't complete well-punctuated sentences, spaCy isn't used to it. 

In your case it seems like the easiest thing would be to just treat each line as its own doc, skipping sentencizing completely. (That said I'm not sure why using `\n` as a punct character didn't work, I'll have to look into that.)
